### ISSUE TITLE
Should gist:Template be a subclass of gist:Specification?

### ISSUE THREAD

**Issue by uscholdm** [semanticarts/gist#60]
A template for something is very much like a specification (in the English sense of the word) for something in that both strongly influence what the nature of the something is or can be.  Thus, perhaps gist:Template should be a subclass of gist:Specification.

**Comment by uscholdm** [comment]
FROM DAVE:
In my mind the difference is the spec establishes acceptance criteria.  That is you can use them to determine where the service "complied" independent of how it was done.  The taskTemlate/template task is about genericity: that is how can you create a set of instances that might satisfy the spec.  

It's easy to imagine generating a set of tasks from a template that's when executed fails to achieve the spec.  

Let's say the spec was 95% up time, and the template task was " generate a help desk ticket when someone complains" 

That template may, or, may not, satisfy the spec

**Comment by uscholdm** [comment]
The text definition of gist:Specification currently is: "A set of requirements to be satisfied by a material, design, product, or service."  It is a subclass of gist:Requirement which is a subclass of gist:Intention.  

Broadly speaking, requirements and specifications come in many different flavors.  Some   dictate what the shape/structure  of something must be (a form to fill in) others dictate performance criteria (95% up time), still others dictate state criteria (acidity level).  A specification may include any or all of these kinds of requirement.  

A form is a template which can be viewed as a specification _requiring_ the filled document be of a certain shape/structure.    While we would not normally say that the form is a [set of] requirement[s] for the document, it does in fact dictate a whole lot about what the document is _required_ to be like. 

Long before gist had a class for Specification, I created one and gave it a more inclusive definition than the current one; Template was a subclass (see below). When Dave created an 'official' gist:Specification some years later, he chose a narrower interpretation, and Template was not a subclass.  

I think it is a legitimate option to broaden the skos:definition of gist:Specification to include a wider range of things that are reasonably thought of as specifications (such as Templates).  We can just change the English and cause minimal impact. 

If we leave it as it is, then we fail to capture the similarity between Template and Specification.



MY ORIGINAL DEFINITIONS from a few years ago.

Here was my original inclusive definition, created with great care and thoughtfulness :
"A body of information specifying what properties, structure and/or results something should or must have after it has been built, fulfilled or completed. The thing may be physical (a building), organizational (a management structure), computational (a program), job or task related (e.g., a CEO, Chief Judge) or changing oil in a car. "  

The define of Template (a subclass of Specification) was:
"A structure that is used as the basis for generating new individuals that have that structure, e.g., a form. A filled-in form has the structure of the form with data entered into some or all of the fields."

**Comment by rjyounes** [comment]
@uscholdm to prepare brief presentation for discussion on January 30.

**Comment by rjyounes** [comment]
A Template is used to make something that follows a specification.

**Comment by uscholdm** [comment]
I agree, but see my long comment above for the history of the term 'Specification' in gist.

Those definitions can be found here:
S:\ClientsAndPartners\Sentara\_Enterprise_Ontology\EntOnt\Version1.0\Final Deliverables\OWL Files\upper1.0.owl

The definition for Specification probably evolved a bit in subsequent projects at Broadridge and Schneider.

**Comment by uscholdm** [comment]
## Summary of group discussion on 12-Nov-20

### Is the meaning of gist:Specification too narrow?
- Consensus was emerging to expand the meaning of gist:Specification beyond acceptance criteria, which is a common use of the term in computing (e.g. a program specification).  This would include the specification of how to do something, which would include things like plan, process, method, recipe.  

- A place to start is the following definition from a prior gist 
  - "A body of information specifying what properties, structure and/or results something should or must have after it has been built, fulfilled or completed. The thing may be physical (a building), organizational (a management structure), computational (a program), job or task related (e.g., a CEO, Chief Judge) or changing oil in a car."

- The question arises (but was not discussed at any length) as to whether there should be two main subclasses, as follows. 
  - one for the acceptance criteria meaning which focusses on *what* something is 
  - the other has a focus on *how* to do something.   Cyc used the term `normative specification' for this.

- There were mixed views on whether gist:Template should be a subclass of gist:Specification. Below are two definitions of gist:Template. The first is current, the second is a former definition.
  - "Something used to make instances in its own image.  In manufacturing this would be specialized as a die to make stamped parts, cookie cutters are templates for cookies, and forms are templates for data."
  - "A structure that is used as the basis for generating new individuals that have that structure, e.g., a form. A filled-in form has the structure of the form with data entered into some or all of the fields."

### Is a Template itself a Spefication?

- On whether Template should be a subclass of Specification, it was noticed that a template is often tool that is used to create something that meets a specification. This would be true of a manufacturing die or cookie cutter.  In those cases, there may well be a specification in the form of a CAD drawing. But some templates could be fashioned directly from ones imagnation; there may not be any specification that came before the template.   If there was, then the die and cookie cutter would be physical embodiments of the specifications.  The question arises:
  - Is a die (or cookie cutter) itself a specification, or just the physical embodiment of a specification.
  - If it just the physical embodiment of the specification, and not a specification, then Template cannot be a subclass of Specification.
  - Would we want to have two individuals: one the specification and one the embodiement that is linked to the specification?
  - This reminds me of the fact that a statue can cease to exist after it is melted down, but the metal still does exist. So strictly, speaking, there must be two separate individuals. But in practice do we really care about that distinction enough to want both?

- This raises the question as to whether a Template is itself a Specification, or just the embodiment of a specification that may or may not exist.

**Comment by rjyounes** [comment]
@uscholdm Thanks for the great summary of our length discussion yesterday. I had a thought about the discussion about Template subclassing Specification: suppose the cookie cutter gets dented. We don't say the _specification_ is defective, we say the template _no longer meets_ the specification. This bolsters the idea that a Template is not a Specification, but a tool for implementing the Specification.

This issue has undergone significant scope creep. The original issue was:

1. Should Template be a subclass of Specification?

We have added an additional 3 issues which are related but not inseparable:
2. Should the definition of Specification be expanded to include process specifications?
3. Should Template be an orphan or assigned to some other superclass? (Since Template currently _is_ an orphan, this is a new issue.)
4. What are the proper relationships between Template, Specification, and Artifact? (and can we prevent Artifact from becoming an orphanage?)

I propose that we deal with the original issue here, and create separate issues for the others: 2 should be its own issue, and 3 and 4 should be rolled into a single issue. Granted that our original issue relates closely to 3 and 4, it can be decided independently, before tackling the other subclassing issues as a whole.

@uscholdm and others involved in yesterday's discussion - @Jamie-SA @marksem @sa-bpelakh - let me know if you're in agreement and I'll create the two additional issues.

My view on 1 is that Template is a tool to implement a Specification rather than a Specification, so it should not be a subclass, and that as part of addressing the current issue, we should modify the human-readable definition accordingly.

Trying to keep these issues to bite-sized chunks as much as possible, I would also like to propose yet another issue to add a restriction to Template requiring it to relate to a Specification in some way. I'm not sure what the predicate would be, and we might need a new one, something like conformsTo or implements. I think I had

**Comment by uscholdm** [comment]
> I propose that we deal with the original issue here, and create separate issues for the others

Yes, this has morphed into what _can_ be different issues, but they are highly interdependent.  For example, the discussion about whether template is a subclass may be impacted by whether we broaden the meaning of specification.   I think we need to clarify and agree on one or more things first before we start creating separate issues. The trick is to identify the dependencies and start in the right place - i.e. meaning less to have to go back and redo if we do them in the wrong order.

> My view on 1 is that Template is a tool to implement a Specification rather than a Specification

This makes ontological/intuitive/philosophical sense, but it does not in principle rule out the possibility that some templates are themselves most usefully thought of and treated as specifications.  It may mean that we should not have them be a subclass, however.

> I would also like to propose yet another issue to add a restriction to Template requiring it to relate to a Specification in some way.

Not yet. This is very closely related to whether a Specification is a superclass of Template. How important is it in practice to separate the specification from the embodiment of the specification? How often would we really want both?  A form is a template. What specification it is a tool to achieve? I'm not convinced it adds any value.  Analogously, how important is it so separate the metal that the statue is made of from the statue?

**Comment by rjyounes** [comment]
If a form is a template, then the definition of template must be changed. Currently it is "Something used to make instances in its own image." A form does not make instances in its own image; it makes data, and that data is not stored as a form but in a structured database of some sort. Either we change the definition or remove the example.

I agree that we should not break this up into different issues unless we are clear on what those issues are. It seems in fact that the first thing to solve is whether the definition of Specification should be expanded to include process specifications. I still recommend that we make an issue for that, with a pointer to this issue, and clarify that this issue cannot be addressed until that one is solved.

The comment about the restriction presupposed that Template is _not_ a subclass of Specification, which of course hasn't been decided yet. 

I'd be interested to hear your views on the dented cookie cutter example. Here's a similar one in case you didn't find that compelling: I'm a baker, and a customer orders a batch of cookies from me to be shaped like maple leaves - this is  the specification. I have a cookie cutter - a template - and start making cookies with it. This cookie cutter breaks midway through, so I make the remainder with a different cookie cutter, which perhaps represents leaves of a different variety of maple tree. The templates have changed but the specification has not. The customer comes in to pick up the cookies and is crestfallen. "Oh, I guess I should have mentioned that I wanted _red_ maple leaves, not _sugar_ maple leaves." The specification has become more precise, but neither of the templates - the cookie cutters - has changed shape. It simply turns out that only one of them met the specification, once the latter was clarified.

**Comment by marksem** [comment]
Candidate text definition for Template:  
Definition "Something that serves as a model for new instances to copy"
Example "A shaped piece of metal, wood, card, or plastic used as pattern; in computing a preset format for a document, a form to fill out."

**Comment by rjyounes** [comment]
DECISION: Specification should include process specifications.

**Comment by uscholdm** [comment]
Interim DECISION: 
- Create separate issue for broadening the definition of Specification (See: https://github.com/semanticarts/gist/issues/405) 
- Reconsider the other points afterwards.

**Comment by marksem** [comment]
I do not think Template is sub of Specification.  Specification is a type of requirement.  I do not see a Template as a type of Requirement.  My Christmas cookie cutter is not in all cases a requirement. It does "serve as a pattern or model for new instances to copy", but it's not automatically a requirement.

**Comment by rjyounes** [comment]
I agree with @marksem and think we should close this issue as a won't fix.

**Comment by rjyounes** [comment]
Closing - won't fix.
